libs/cardano-ledger-core/src/Cardano/Ledger/UMapCompact.hs

```diff
@@ -0,0 +1,733 @@
+notMember k um = not (member k um)
+
+-- =====================================================
+
+-- eval (dom rewards' ◁ iRReserves (_irwd ds) :: RewardAccounts (Crypto era))
+-- eval (dom rewards' ◁ iRTreasury (_irwd ds) :: RewardAccounts (Crypto era))
+
+(◁) :: View c k v -> Map k u -> Map k u
+(◁) = domRestrict
+
+domRestrict :: View c k v -> Map k u -> Map k u
+domRestrict (Rewards (UMap tmap _)) m = intersectDomPLeft p m tmap
+  where
+    p _ (Triple (SJust _) _ _) = True
+    p _ _ = False
+domRestrict (Delegations (UMap tmap _)) m = intersectDomPLeft p m tmap
+  where
+    p _ (Triple _ _ (SJust _)) = True
+    p _ _ = False
+domRestrict (Ptrs (UMap _ pmap)) m = Map.intersection m pmap
+
+-- ==========================
+
+instance
+  (Crypto c) =>
+  ToCBOR (Trip c)
+  where
+  toCBOR (Triple coin ptr pool) =
+    encodeListLen 3 <> toCBOR coin <> toCBOR ptr <> toCBOR pool
+
+instance Crypto c => FromSharedCBOR (Trip c) where
+  type Share (Trip c) = Interns (KeyHash 'StakePool c)
+  fromSharedCBOR is =
+    decodeRecordNamed "Triple" (const 3) $
+      do
+        a <- fromCBOR
+        b <- fromCBOR
+        c <- fromShareCBORfunctor is
+        pure (Triple a b c)
+
+instance Crypto c => ToCBOR (UMap c) where
+  toCBOR (UMap tripmap ptrmap) =
+    encodeListLen 2 <> encodeMap toCBOR toCBOR tripmap <> encodeMap toCBOR toCBOR ptrmap
+
+instance Crypto c => FromSharedCBOR (UMap c) where
+  type
+    Share (UMap c) =
+      (Interns (Credential 'Staking c), Interns (KeyHash 'StakePool c))
+  fromSharedPlusCBOR =
+    StateT
+      ( \(a, b) ->
+          decodeRecordNamed "UMap" (const 2) $ do
+            tripmap <- decodeMap (interns a <$> fromCBOR) (fromSharedCBOR b)
+            let a' = internsFromMap tripmap <> a
+            ptrmap <- decodeMap fromCBOR (interns a' <$> fromCBOR)
+            pure (UMap tripmap ptrmap, (a', b))
+      )
+
+-- ==================================================
+-- derived operations
+
+-- DFUNS
+
+-- | Find the value associated with a key from a View, return the default if the key is not there.
+findWithDefault :: a -> k -> View c k a -> a
+findWithDefault d k = fromMaybe d . lookup k
+
+-- | A View is a view, so the size of the view is NOT the same as the size of
+-- the underlying triple map.
+size :: View c k a -> Int
+size (Ptrs (UMap _ ptrmap)) = Map.size ptrmap
+size x = foldl' (\count _v -> count + 1) 0 x
+
+-- | Create a UMap from 3 separate maps. For use in tests only.
+unify ::
+  Map (Credential 'Staking c) Coin ->
+  Map (Credential 'Staking c) (KeyHash 'StakePool c) ->
+  Map Ptr (Credential 'Staking c) ->
+  UMap c
+unify rews dels ptrss = um3
+  where
+    um1 = unView $ Map.foldlWithKey' (\um k v -> insert' k (compactCoinOrError v) um) (Rewards empty) rews
+    um2 = unView $ Map.foldlWithKey' (\um k v -> insert' k v um) (Delegations um1) dels
+    um3 = unView $ Map.foldlWithKey' (\um k v -> insert' k v um) (Ptrs um2) ptrss
+
+compactCoinOrError :: Coin -> CompactForm Coin
```

Comment:
Let's add a call stack, otherwise if it does happen, we will not know in which place it was actually triggered.
```suggestion
compactCoinOrError :: HasCallStack => Coin -> CompactForm Coin
```

Reply:
done

Reply:
done